app/amp_start.py

```diff
@@ -13,11 +13,25 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from start import Handler
+import urlparse
+import start
+import utils
 
 
-class AMPstartHandler(Handler):
-    repo_required = False
+class Handler(start.Handler):
 
     def get(self):
+        self.env.amp = True
+        self.env.canonical_url = self.set_canonical_url_param(self.request.url,
```

Comment:
Changed. First, I thought put a language params to the URL is better, but I realized AMP page never knows language settings before user coming.

Reply:
That's a good point. It's possible that the user access /amp_start?lang=ja where we know the language, right? You can write self.get_url('/', lang=self.env.lang) to support that case.

Reply:
Another thing I want to discuss is that should I put the same logic on start.py? I'm wondering most user's default is English. Probably, we keep just in here.

Reply:
Are you saying that there are two options:

Option 1:

```python
# In amp_start.py:
self.env.canonical_url = self.get_url('/', lang=self.env.lang)
# In start.py:
self.env.amp_url = self.get_url('/amp_start')
```

Option 2:

```python
# In amp_start.py:
self.env.canonical_url = self.get_url('/', lang=self.env.lang)
# In start.py:
self.env.amp_url = self.get_url('/amp_start', lang=self.env.lang)
```

And you prefer Option 2? What would change among these options, and why do you think Option 2 is better?

Reply:
My thought is AMP pages caching by Google and Search Engine shows the page which indexed. If passible to show the amp_start + lang on each language (en, ja, cn,,,) Option 2 is better, but I have no idea how to show the English search engine to show en params. I read the AMP pages there is no i18n options in there yet.

Reply:
Looking at Google's search result:
https://www.google.co.jp/search?safe=active&q=+site:google.org+person+finder&sa=X&ved=0ahUKEwiDlIHv94_XAhVPhrwKHdbeDLwQrAMIMg&biw=1569&bih=875
Looks like they often index pages without lang param, but sometimes they do:
https://www.google.org/personfinder/japan
https://www.google.org/personfinder/japan?lang=en
https://www.google.org/personfinder/japan?lang=es-419

So we should be prepared for both cases. Maybe it' the safest to simply preserve the lang param i.e., inherit it if specified but omit it if it's omitted?

You can probably achieve it by writing this way:
```python
# In amp_start.py:
self.env.canonical_url = self.get_url('/', lang=self.params.lang)
# In start.py:
self.env.amp_url = self.get_url('/amp_start', lang=self.params.lang)
```
But I haven't tried. Try it to see if it works.

The trick is to use self.params.lang instead of self.env.lang. e.g., If the lang param is not given in the URL but we guess it's English, self.params.lang should be None but self.env.lang is "en".

Reply:
I tried self.params.lang it working as I intended. It's better than self.env.lang I'll submit once I finished unit tests.